src/completion.ts

```diff
@@ -101,44 +129,80 @@ export function asCompletionItem(
-    optionalReplacementSpan: ts.server.protocol.TextSpan | undefined,
-    position: lsp.Position,
-    document: LspDocument,
-    features: SupportedFeatures,
-): { insert?: lsp.Range; replace: lsp.Range; } | undefined {
-    if (replacementSpan) {
-        // If TS provides an explicit replacement span with an entry, we should use it and not provide an insert.
-        return {
-            replace: ensureRangeIsOnSingleLine(Range.fromTextSpan(replacementSpan), document),
-        };
+function getFilterText(entry: ts.server.protocol.CompletionEntry, wordRange: lsp.Range | undefined, line: string, insertText: string | undefined): string | undefined {
+    // Handle private field completions
+    if (entry.name.startsWith('#')) {
+        const wordStart = wordRange ? line.charAt(wordRange.start.character) : undefined;
+        if (insertText) {
+            if (insertText.startsWith('this.#')) {
+                return wordStart === '#' ? insertText : insertText.replace(/&this\.#/, '');
+            } else {
+                return wordStart;
+            }
+        } else {
+            return wordStart === '#' ? undefined : entry.name.replace(/^#/, '');
+        }
     }
-    if (features.completionInsertReplaceSupport && optionalReplacementSpan) {
-        const range = ensureRangeIsOnSingleLine(Range.fromTextSpan(optionalReplacementSpan), document);
-        return {
-            insert: lsp.Range.create(range.start, position),
-            replace: ensureRangeIsOnSingleLine(range, document),
-        };
+
+    // For `this.` completions, generally don't set the filter text since we don't want them to be overly prioritized. #74164
+    if (insertText?.startsWith('this.')) {
+        return undefined;
     }
+
+    // Handle the case:
+    // ```
+    // const xyz = { 'ab c': 1 };
+    // xyz.ab|
+    // ```
+    // In which case we want to insert a bracket accessor but should use `.abc` as the filter text instead of
+    // the bracketed insert text.
+    if (insertText?.startsWith('[')) {
+        return insertText.replace(/^\[['"](.+)[['"]\]$/, '.$1');
+    }
+
+    // In all other cases, fallback to using the insertText
+    return insertText;
 }
 
-function ensureRangeIsOnSingleLine(range: lsp.Range, document: LspDocument): lsp.Range {
-    if (range.start.line !== range.end.line) {
-        return lsp.Range.create(range.start, document.getLineEnd(range.start.line));
+function getRangeFromReplacementSpan(entry: ts.server.protocol.CompletionEntry, document: LspDocument) {
+    if (!entry.replacementSpan) {
+        return;
+    }
+
+    let replacementSpan = Range.fromTextSpan(entry.replacementSpan);
+
+    if (replacementSpan && replacementSpan.start.line !== replacementSpan.end.line) {
+        replacementSpan = lsp.Range.create(replacementSpan.start, document.getLineEnd(replacementSpan.start.line));
     }
-    return range;
+
+    return {
+        inserting: replacementSpan,
+        replacing: replacementSpan,
+    };
```

Comment:
The previous code has a comment:

```ts
// If TS provides an explicit replacement span with an entry, we should use it and not provide an insert.
```

and also only provided a `textEdit` in this case since it doesn't really make sense to provide both insert & replace if both are the same anyway.

Can we keep this logic from before?